Add Go structs for data model 2.0

spa-raj opened the pull request:
## What does this PR change?
* Create Go model structs in core/pkg/model/cluster.go based on proto message definitions from protos/kubemodel/*.proto for the data model 2.0 project. These structs will replace proto-generated code when transitioning to bingen or JSON serialization.

Implemented structs from proto definitions:
 - Provider and ControllerKind enums (without unspecified values)
 - Window, Resolution types for time-based aggregation
 - Container with full metrics and diagnostic support
 - Controller for Kubernetes workload controllers
 - DiagnosticResult matching core/pkg/diagnostics/diagnostics.go
 - GPUDevice with sharing, usage metrics, and memory tracking
 - GPUUsage for container-level GPU consumption
 - Namespace, Service, ServicePort with network metrics
 - Node with cost and usage tracking
 - Pod with full resource metrics
 - Volume and PersistentVolumeClaim with storage tracking
 - Cluster as top-level aggregator

Key design decisions:
 - Use core/pkg/diagnostics/diagnostics.go as source of truth for DiagnosticResult (not proto) since proto will be deprecated
 - Include per-resource diagnostic fields, exclude cluster-level diagnostics list (not in proto schema)
 - Native Go types (time.Time, map[string]any) instead of proto types
 - Clean field naming without proto legacy duplicates

These native Go structs form the foundation for data model 2.0 serialization via bingen or JSON, removing dependency on protobuf generated code.

## Does this PR relate to any other PRs?
* Based on the proto definitions from #3369 and #3406 

## How will this PR impact users?
* No immediate impact to users. This PR is the foundation for the Data model 2.0 project under the Linux foundation mentorship program.

## Does this PR address any GitHub or Zendesk issues?
* No

## How was this PR tested?
1. **Compilation verification**: Built core/pkg/model package successfully to
     ensure all struct definitions are syntactically correct.

2. **Proto alignment validation**: Cross-referenced each Go struct against
     corresponding proto messages in protos/kubemodel/*.proto to verify all 
     fields are present, correctly typed, and no duplicates exist.

## Does this PR require changes to documentation?
* No

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* No, this PR contains breaking changes.

spa-raj committed:
Add Go structs for data model 2.0 from proto definitions

Create Go model structs in core/pkg/model/cluster.go based on proto
message definitions from protos/kubemodel/*.proto for the data model 2.0
project. These structs will replace proto-generated code when transitioning
to bingen or JSON serialization.

Implemented structs from proto definitions:
 - Provider and ControllerKind enums (without unspecified values)
 - Window, Resolution types for time-based aggregation
 - Container with full metrics and diagnostic support
 - Controller for Kubernetes workload controllers
 - DiagnosticResult matching core/pkg/diagnostics/diagnostics.go
 - GPUDevice with sharing, usage metrics, and memory tracking
 - GPUUsage for container-level GPU consumption
 - Namespace, Service, ServicePort with network metrics
 - Node with cost and usage tracking
 - Pod with full resource metrics
 - Volume and PersistentVolumeClaim with storage tracking
 - Cluster as top-level aggregator

Key design decisions:
 - Use core/pkg/diagnostics/diagnostics.go as source of truth for
   DiagnosticResult (not proto) since proto will be deprecated
 - Include per-resource diagnostic fields, exclude cluster-level
   diagnostics list (not in proto schema)
 - Native Go types (time.Time, map[string]any) instead of proto types
 - Clean field naming without proto legacy duplicates

These native Go structs form the foundation for data model 2.0
serialization via bingen or JSON, removing dependency on protobuf
generated code.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj commented:
@nikovacevic @Sean-Holcomb @ameijer Please review this PR.

Sean-Holcomb committed:
Merge branch 'develop' into feature/kubemodel-go-structs

Sean-Holcomb commented:
Thanks for pivoting this over, we will get some comments on this later

ameijer commented on core/pkg/model/cluster.go:
```diff
@@ -0,0 +1,268 @@
+	RamByteHours           int64             `json:"ramByteHours"`
+	RamBytesRequestAverage int64             `json:"ramBytesRequestAverage"`
+	RamBytesUsageAverage   int64             `json:"ramBytesUsageAverage"`
+	RamBytesUsageMax       int64             `json:"ramBytesUsageMax"`
```
do we want to break this out to a container, then have a pod hold n many containers?

ameijer replied:
oh, never mind, I found container. I don't think we would need these fields here then?

spa-raj replied:
Pods can have multiple containers, and these pod-level fields provide pre-aggregated metrics across all containers, eliminating the need for clients to manually sum container metrics. We can remove them if we want the model to be lightweight.

aaliomer-ibm commented on core/pkg/model/cluster.go:
```diff
@@ -0,0 +1,268 @@
+	Name                   string            `json:"name"`
+	CreationTime           *time.Time        `json:"creationTime,omitempty"`
+	DeletionTime           *time.Time        `json:"deletionTime,omitempty"`
+	CpuCoreHours           float32           `json:"cpuCoreHours"`
```
Lot of those float32 should probably be float64

aaliomer-ibm commented on core/pkg/model/cluster.go:
```diff
@@ -0,0 +1,268 @@
+	CpuCoreRequestAverage  float32           `json:"cpuCoreRequestAverage"`
+	CpuCoreUsageAverage    float32           `json:"cpuCoreUsageAverage"`
+	CpuCoreUsageMax        float32           `json:"cpuCoreUsageMax"`
+	RamByteHours           int64             `json:"ramByteHours"`
```
use uint64

aaliomer-ibm commented on core/pkg/model/cluster.go:
```diff
@@ -0,0 +1,268 @@
+// ServicePort represents a service port
+type ServicePort struct {
+	Name       string `json:"name"`
+	Port       int32  `json:"port"`
```
uint16

aaliomer-ibm commented on core/pkg/model/cluster.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+// Container represents a container within a pod (allocated resource)
+type Container struct {
```
is it possible to attribute the volume storage / file system usage here?

aaliomer-ibm commented on core/pkg/model/cluster.go:
```diff
@@ -0,0 +1,268 @@
+	Window   *Window  `json:"window,omitempty"`
+
+	// Resources within the cluster
+	Containers             []Container             `json:"containers,omitempty"`
```
the model is very flat where the object structure is hierarchal

aaliomer-ibm commented on core/pkg/model/cluster.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+// GPUUsage represents GPU usage metrics
+type GPUUsage struct {
```
shouldn't this be part of a pod or a container?

spa-raj replied:
Added in container

spa-raj committed:
Refactor cluster model: improve numeric types, nest GPU usage in containers, and add ephemeral storage metrics

Changes float32 to float64 for precision, uses uint64 for bytes/counts, moves GpuUsages into Container struct for proper ownership, and adds container-level ephemeral storage
fields for consistency with CPU/RAM metrics.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Refactor kubemodel: extract structs into separate files and move resources to KubeModelSet

- Extract all struct definitions from cluster.go into individual files:
   - provider.go, resolution.go, controller.go (with enums)
   - window.go, diagnostic.go
   - gpu_usage.go, gpu_device.go
   - container.go, namespace.go, service.go, node.go, pod.go
   - volume.go, pvc.go (using abbreviated naming)

- Move resource collections from Cluster struct to KubeModelSet as maps
  - Cluster now only contains cluster-level metadata
  - All resources (Containers, Controllers, GpuDevices, Namespaces, etc.)
    are now stored as maps in KubeModelSet for efficient lookups

- Implement namespace name-to-ID indexing
  - Add GetNamespaceByName() method for O(1) lookups by name
  - Update RegisterNamespace() to maintain the index

- Improve KubeModelSet initialization and methods
  - Fix NewKubeModelSet to use Resolution instead of end time
  - Implement IsEmpty() to check for nil, missing cluster, or no resources
  - Fix RegisterResourceQuota parameter handling

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Replace Window resolution with explicit end time for dynamic time boundaries

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

nikovacevic commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -0,0 +1,9 @@
+
+// Cluster represents the top-level Kubernetes cluster
+type Cluster struct {
+	ID       string   `json:"id"`
+	Provider Provider `json:"provider"`
+	Account  string   `json:"account"`
+	Name     string   `json:"name"`
```
Need to decide where the data comes from for `ID` and `Name`. Cloudy uses `kube-system` namespace's `UID` as the cluster `ID`.

nikovacevic commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,27 @@
+	Name                       string            `json:"name"`
+	CreationTime               *time.Time        `json:"creationTime,omitempty"`
+	DeletionTime               *time.Time        `json:"deletionTime,omitempty"`
+	CpuCoreHours               float64           `json:"cpuCoreHours"`
```
What unit for CPU?

nikovacevic replied:
mCPU * sec

nikovacevic replied:
`uint64`

nikovacevic commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,27 @@
+	StorageBytesRequestAverage uint64            `json:"storageBytesRequestAverage"`
+	StorageBytesUsageAverage   uint64            `json:"storageBytesUsageAverage"`
+	StorageBytesUsageMax       uint64            `json:"storageBytesUsageMax"`
+	GpuUsages                  []GPUUsage        `json:"gpuUsages,omitempty"`
```
We need to deal with attached devices (GPUs, storage, etc.) -- something map-based.

Allow for things to evolve into the future.

nikovacevic commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,27 @@
+	PodID                      string            `json:"podId"`
+	Name                       string            `json:"name"`
+	CreationTime               *time.Time        `json:"creationTime,omitempty"`
+	DeletionTime               *time.Time        `json:"deletionTime,omitempty"`
```
Prefer (Start, End) within the Window

nikovacevic commented on core/pkg/model/kubemodel/diagnostic.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+// DiagnosticResult represents the result of a diagnostic run
+type DiagnosticResult struct {
```
Should Diagnostics live at the Object level or the Set level?

nikovacevic reviewed (approved):
Partial review with commentary from 11/4 meeting

nikovacevic reviewed (changes requested):
Let's make changes in accord with previous comment

spa-raj committed:
Update kubemodel with MVP device architecture

   - Replace GPU-specific arrays with generic map-based device model
   - Add Device, DeviceType, and DeviceUsage structs
   - Update Container to use Devices map[string]*DeviceUsage
   - Update Container time fields: CreationTime/DeletionTime -> Start/End
   - Update KubeModelSet to use Devices map[string]*Device
   - Remove old gpu_device.go and gpu_usage.go files
   - Update CPU/RAM/Storage units to seconds and uint64 for consistency
   - Focus on OpenCost-compatible metrics for MVP

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Merge branch 'opencost:develop' into feature/kubemodel-go-structs

spa-raj committed:
Replace CreationTime/DeletionTime with Start/End in kubemodel structs

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>
